### PR TITLE
swt: 4.5 -> 4.33 & more

### PR DESCRIPTION
**Pull request by doronbehar**
## Things done

Some seem to not use the swt shared libraries or the `swt.jar`.

- **dbeaver-bin: don't depend on swt (not needed at all)**
- **portfolio: don't depend on swt (not needed at all)**
- **ipscan: mark as broken (doesn't really launch)**
- **vuze: mark as broken**

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gepbird** [comment]
swt (and glib and gtk3) was added to dbeaver-bin because of https://github.com/NixOS/nixpkgs/pull/335633.
TLDR: without those changes it works for some people but not for others.
I'll test if it works with this PR using docker where the issue can be easily reproduced.

**Comment by Shawn8901** [comment]
Portfolio had the same crash, so I added set to unbreak it. I can test later if portfolio starts with the said changes.

**Comment by gepbird** [comment]
Tested dbeaver-bin on this PR and works well, thanks for the cleanup!

**Review by gepbird — APPROVED** [review]

**Comment by Shawn8901** [comment]
portfolio also still works fine with the pr.

**Comment by doronbehar** [comment]
Thanks for your approvals, I'll wait a few days for @ardumont to approve the changes for `tuxguitar`, and merge this next week or so.

**Comment by lopsided98** [comment]
I was trying to use this to make DataExplorer work (its bundled SWT is broken, same error as https://github.com/NixOS/nixpkgs/issues/340048), and it looks like there are some additional resource files that must be copied to the JAR to make everything work correctly. It fails to start with:
```
Exception in thread "main" java.lang.ExceptionInInitializerError
Caused by: java.lang.NullPointerException
	at java.base/java.io.Reader.<init>(Reader.java:168)
	at java.base/java.io.InputStreamReader.<init>(InputStreamReader.java:76)
	at org.eclipse.swt.graphics.Device.lambda$overrideThemeValues$0(Device.java:763)
	at org.eclipse.swt.graphics.Device.overrideThemeValues(Device.java:778)
	at org.eclipse.swt.graphics.Device.init(Device.java:737)
	at org.eclipse.swt.widgets.Display.init(Display.java:3476)
	at org.eclipse.swt.graphics.Device.<init>(Device.java:168)
	at org.eclipse.swt.widgets.Display.<init>(Display.java:633)
	at org.eclipse.swt.widgets.Display.<init>(Display.java:624)
	at org.eclipse.swt.widgets.Display.getDefault(Display.java:2393)
	at gde.GDE.<clinit>(GDE.java:453)
```

This happens because it tries to load `/org/eclipse/swt/internal/gtk/swt_functional_gtk_3_20.css`, which is missing from the JAR. There are also `.properties` translation files that are missing.

I added this command to `postBuild` and then it worked correctly:
```
find org/ -not -name '*.java' -not -name '*.html' -type f -exec cp {} out/{} \;
```

**Comment by doronbehar** [comment]
Thanks @lopsided98 for commenting! The error you show seems different from what we observe in https://github.com/NixOS/nixpkgs/issues/340048 . And unfortunately your solution did not help for `hdfview`... I won't mind commit that anyway. Feel free to push further changes to DataExplorer if you want, or let me know if I did it correctly.

**Comment by lopsided98** [comment]
I get the same error as https://github.com/NixOS/nixpkgs/issues/340048 when I try to use the bundled SWT that comes with DataExplorer. When I replaced it with the version from nixpkgs, I got the error I posted.

**Comment by doronbehar** [comment]
Oh I see. I get the same error with and without that change to swt, while I apply the following patch upon hdfview's expression:

```patch
commit f5a7868f9a72b5fa3acfb757d0349e9e7e1bfc50
Author: Doron Behar <doron.behar@gmail.com>
Date:   Sun Oct 6 22:35:50 2024 +0300

    hdfview: unbreak

diff --git a/pkgs/tools/misc/hdfview/default.nix b/pkgs/tools/misc/hdfview/default.nix
index 3274762d87c3..48656b5a1b2b 100644
--- a/pkgs/tools/misc/hdfview/default.nix
+++ b/pkgs/tools/misc/hdfview/default.nix
@@ -6,6 +6,7 @@
   jdk,
   hdf4,
   hdf5,
+  swt,
   makeDesktopItem,
   copyDesktopItems,
   strip-nondeterminism,
@@ -46,9 +47,12 @@ stdenv.mkDerivation (finalAttrs: {
   buildPhase =
     let
       arch = if stdenv.hostPlatform.isx86_64 then "x86_64" else "aarch64";
+      os = if stdenv.hostPlatform.isDarwin then "osx" else "linux";
     in
     ''
       runHook preBuild
+      # https://forum.hdfgroup.org/t/building-hdfview-3-1-0-on-centos-6-swt-library-not-found/5698/4
+      cp ${swt}/jars/swt.jar lib/ext/swt/${os}/${arch}/swt.jar
 
       ant createJPackage -Dmachine.arch=${arch}
 
@@ -102,12 +106,5 @@ stdenv.mkDerivation (finalAttrs: {
     platforms = lib.platforms.unix;
     maintainers = with lib.maintainers; [ jiegec ];
     mainProgram = "HDFView";
-    # Startup issue is described here:
-    # https://github.com/NixOS/nixpkgs/issues/340048 A possible solution is
-    # suggested here:
-    # https://forum.hdfgroup.org/t/building-hdfview-3-1-0-on-centos-6-swt-library-not-found/5698
-    # But it requires us to update swt, which is a bit hard, the swt update is tracked here:
-    # https://github.com/NixOS/nixpkgs/issues/219771
-    broken = true;
   };
 })
```

As we are not sure if it actually helps, I kept that commit one of my local branch, and removed it from this PR.

**Comment by lopsided98** [comment]
IMO it is clearly wrong that the those `.css` and `.properties` files are missing from the SWT JAR, regardless of any other issues.

Also, for reference, these are the changes I made to get DataExplorer working:
```
diff --git a/pkgs/by-name/da/dataexplorer/package.nix b/pkgs/by-name/da/dataexplorer/package.nix
index 4ee2b0a0e81a..09406b6b3beb 100644
--- a/pkgs/by-name/da/dataexplorer/package.nix
+++ b/pkgs/by-name/da/dataexplorer/package.nix
@@ -4,11 +4,12 @@
 , ant
 # executable fails to start for jdk > 17
 , jdk17
+, swt
 , makeWrapper
 , strip-nondeterminism
-}:
-
-stdenv.mkDerivation (finalAttrs: {
+}: let
+  swt-jdk17 = swt.override { jdk = jdk17; };
+in stdenv.mkDerivation (finalAttrs: {
   pname = "dataexplorer";
   version = "3.9.0";
 
@@ -40,12 +41,15 @@ stdenv.mkDerivation (finalAttrs: {
     runHook preInstall
 
     ant -Dprefix=$out/share/ -f build/build.xml install
+    # Use SWT from nixpkgs
+    ln -sf '${swt-jdk17}/jars/swt.jar' "$out/share/DataExplorer/java/ext/swt.jar"
 
     # The sources contain a wrapper script in $out/share/DataExplorer/DataExplorer
     # but it hardcodes bash shebang and does not pin the java path.
     # So we create our own wrapper, using similar cmdline args as upstream.
     mkdir -p $out/bin
     makeWrapper ${jdk17}/bin/java $out/bin/DataExplorer \
+      --prefix LD_LIBRARY_PATH : '${swt-jdk17}/lib' \
       --add-flags "-Xms64m -Xmx3092m -jar $out/share/DataExplorer/DataExplorer.jar" \
       --set SWT_GTK3 0
 
diff --git a/pkgs/by-name/sw/swt/package.nix b/pkgs/by-name/sw/swt/package.nix
index 04d880dc3368..b06eaa9ab08c 100644
--- a/pkgs/by-name/sw/swt/package.nix
+++ b/pkgs/by-name/sw/swt/package.nix
@@ -75,7 +75,9 @@ stdenv.mkDerivation (finalAttrs: {
   postBuild = ''
     cd ../
     mkdir out
-    find org/ -name '*.java' -type f -exec javac -d out/ {} +
+    find org/ -name '*.java' -type f -exec javac -encoding utf8 -d out/ {} +
+    # Copy non Java resource files
+    find org/ -not -name '*.java' -not -name '*.html' -type f -exec cp {} out/{} \;
   '';
 
   # The makefile doesn't have an install target, the installation of the shared

```

**Comment by doronbehar** [comment]
OK great! I'm working now on a similar fix for hdfview.. Pushed your changes in the meantime.

**Comment by doronbehar** [comment]
I'm merging as there are no objections for almost a week, and for sure this improves the situation.
